pkg/interop/client.go

```diff
@@ -0,0 +1,328 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interop
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"go.thethings.network/lorawan-stack/pkg/config"
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+const (
+	// loRaAllianceDomain is the domain of LoRa Alliance.
+	loRaAllianceDomain = "lora-alliance.org"
+
+	// LoRaAllianceJoinEUIDomain is the LoRa Alliance domain used for JoinEUI resolution.
+	LoRaAllianceJoinEUIDomain = "joineuis." + loRaAllianceDomain
+
+	// LoRaAllianceNetIDDomain is the LoRa Alliance domain used for NetID resolution.
+	LoRaAllianceNetIDDomain = "netids." + loRaAllianceDomain
+
+	defaultHTTPSPort = 443
+)
+
+func serverURL(scheme, fqdn, path string, port uint32) string {
+	if scheme == "" {
+		scheme = "https"
+	}
+	if port == 0 {
+		port = defaultHTTPSPort
+	}
+	if path != "" {
+		path = fmt.Sprintf("/%s", path)
+	}
+	return fmt.Sprintf("%s://%s:%d%s", scheme, fqdn, port, path)
+}
+
+func newHTTPRequest(url string, pld interface{}, headers map[string]string) (*http.Request, error) {
+	buf := &bytes.Buffer{}
+	if err := json.NewEncoder(buf).Encode(pld); err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest(http.MethodPost, url, buf)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	for k, v := range headers {
+		req.Header.Set(k, v)
+	}
+	return req, nil
+}
+
+func JoinServerFQDN(eui types.EUI64, domain string) string {
+	o := []rune(strings.ToLower(eui.String()))
+	r := make([]rune, 2*len(o)-1)
+	// Inspired by https://github.com/golang/example/blob/46695d81d1fae905a270fb7db8a4d11a334562fe/stringutil/reverse.go#L20-L27
+	for i, j := 0, len(o)-1; i < len(o)/2; i, j = i+1, j-1 {
+		r[i+i], r[j+j] = o[j], o[i]
+		r[i+i+1] = '.'
+		r[j+j-1] = '.'
+	}
+	if domain == "" {
+		domain = LoRaAllianceJoinEUIDomain
+	}
+	return fmt.Sprintf("%s.%s", string(r), domain)
```

Comment:
Something like

```go
fmt.Sprintf(
	"%02x.%02x.%02x.%02x.%02x.%02x.%02x.%02x.%s", 
	eui[7], eui[6], eui[5], eui[4], eui[3], eui[2], eui[1], eui[0],
	domain,
)
```

would also work, and then we'd actually understand what's going on

Reply:
Your suggestion produces an invalid result, but I made it work with:
```
	return fmt.Sprintf(
		"%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%01x.%s",
		eui[7]&0xf, eui[7]>>4,
		eui[6]&0xf, eui[6]>>4,
		eui[5]&0xf, eui[5]>>4,
		eui[4]&0xf, eui[4]>>4,
		eui[3]&0xf, eui[3]>>4,
		eui[2]&0xf, eui[2]>>4,
		eui[1]&0xf, eui[1]>>4,
		eui[0]&0xf, eui[0]>>4,
		domain,
	)
```

Reply:
Ah yeah of course, they were by nibble, not by byte. All the more reason to make this explicit. 👍